src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -607,22 +607,22 @@ std::pair<bool, Uptane::Target> SotaUptaneClient::downloadImage(Uptane::Target t
     report_queue->enqueue(std_::make_unique<EcuDownloadStartedReport>(ecu.first, correlation_id));
   }
 
-  bool success = false;
-  const int max_tries = 3;
-  int tries = 0;
-  std::chrono::milliseconds wait(500);
-
   KeyManager keys(storage, config.keymanagerConfig());
   keys.loadKeys();
   auto prog_cb = [this](const Uptane::Target &t, const std::string description, unsigned int progress) {
     report_progress_cb(events_channel.get(), t, description, progress);
   };
 
-  while (tries++ < max_tries) {
+  bool success = false;
+  const int max_tries = 3;
+  int tries = 0;
+  std::chrono::milliseconds wait(500);
+
+  for (; tries < max_tries; tries++) {
     success = package_manager_->fetchTarget(target, *uptane_fetcher, keys, prog_cb, token);
     if (success) {
       break;
-    } else if (tries < max_tries) {
+    } else if (tries < max_tries - 1) {
```

Comment:
So we won't sleep in the last iteration (i = 2)? My guess would be that we don't need an `else if` here.

Reply:
Right. Why bother sleeping if we've tried three times and failed? No need to sleep before moving on to the next thing.

Reply:
Sure. Somehow I always get confused by this simple loop.

Reply:
Me too. This is at least the third time we've tried to get it right despite it being so deceptively simple.